input/pipewire.c

```diff
@@ -202,53 +200,56 @@ void *input_pipewire(void *audiodata) {
         pw_properties_set(props, "channelmix.upmix", "true");
 
-        if (data.cava_audio->channels < 2) {
+        if (data->cava_audio->channels < 2) {
             // N to 1 with all channels shown
             params[0] = spa_format_audio_raw_build(
                 &b, SPA_PARAM_EnumFormat,
-                &SPA_AUDIO_INFO_RAW_INIT(.format = audio_format, .rate = data.cava_audio->rate,
-                                         .channels = data.cava_audio->channels, ));
+                &SPA_AUDIO_INFO_RAW_INIT(.format = audio_format, .rate = data->cava_audio->rate,
+                                         .channels = data->cava_audio->channels, ));
         } else {
             // N to 2 with all channels shown
             params[0] = spa_format_audio_raw_build(
                 &b, SPA_PARAM_EnumFormat,
-                &SPA_AUDIO_INFO_RAW_INIT(.format = audio_format, .rate = data.cava_audio->rate,
-                                         .channels = data.cava_audio->channels,
+                &SPA_AUDIO_INFO_RAW_INIT(.format = audio_format, .rate = data->cava_audio->rate,
+                                         .channels = data->cava_audio->channels,
                                          .position = {SPA_AUDIO_CHANNEL_FL,
                                                       SPA_AUDIO_CHANNEL_FR}, ));
         }
     } else {
         // N to 2 with only FL and FR shown
         params[0] = spa_format_audio_raw_build(
             &b, SPA_PARAM_EnumFormat,
-            &SPA_AUDIO_INFO_RAW_INIT(.format = audio_format, .rate = data.cava_audio->rate,
-                                     .channels = data.cava_audio->channels, ));
+            &SPA_AUDIO_INFO_RAW_INIT(.format = audio_format, .rate = data->cava_audio->rate,
+                                     .channels = data->cava_audio->channels, ));
     }
 
-    data.stream = pw_stream_new_simple(pw_main_loop_get_loop(data.loop), "cava", props,
-                                       &stream_events, &data);
+    data->stream = pw_stream_new_simple(pw_main_loop_get_loop(data->loop), "cava", props,
+                                        &stream_events, data);
 
-    int status = pw_stream_connect(data.stream, PW_DIRECTION_INPUT, PW_ID_ANY,
+    int status = pw_stream_connect(data->stream, PW_DIRECTION_INPUT, PW_ID_ANY,
                                    PW_STREAM_FLAG_AUTOCONNECT | PW_STREAM_FLAG_MAP_BUFFERS |
                                        PW_STREAM_FLAG_RT_PROCESS,
                                    params, 1);
 
     if (status < 0) {
-        data.cava_audio->terminate = 1;
-        sprintf(data.cava_audio->error_message,
+        data->cava_audio->terminate = 1;
+        sprintf(data->cava_audio->error_message,
                 __FILE__ ": Could not connect stream. Is your system running pipewire? Maybe try "
                          "pulse input method instead.");
-        return 0;
+        goto error_free;
```

Comment:
this could be
```
else {
    pw_main_loop_run(data->loop);
    pw_stream_destroy(data->stream);
}
pw_loop_destroy_source(pw_main_loop_get_loop(data->loop), data->timer);
...
```
could it not?

Reply:
Yes, good point.  Let me recheck once I reach PC